### ISSUE TITLE
Connection retries indefinitely when amqp-broker is not available, broker_connection_max_retries not used.

### ISSUE THREAD

**Issue by metteludwig** [celery/celery#4328]
## Checklist

- [N/A] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
- [X] I have verified that the issue exists against the `master` branch of Celery.

## Steps to reproduce
Run the following python-code, _with the amqp-broker disabled_ or shut down. I ran this with a bare virtualenv with latest Celery/billiard/vine/amqp/kombu/pytz as specified in the README.

foo.py:
```
from celery import Celery
app = Celery('foo', broker_url='amqp://', broker_connection_max_retries='2')

@app.task
def task(a, b):
   return a+b

task.delay(1,2)
```

pip freeze:
- amqp==2.2.2
- billiard==3.5.0.3
- celery==4.1.0
- kombu==4.1.0
- pytz==2017.2
- vine==1.1.4

## Expected behavior
Should raise an error after Kombu has tried connecting two times, according to `broker_connection_max_retries`.

## Actual behavior
Kombu retries connecting indefinitely. This behaviour be observed and verified in `kombu.utils.functional.retry_over_time(...)`

It was a real bugger trying to find this culprit, my fix was of course to start the amqp-broker (forgot to run it on our CI-server), but the troubleshooting was tiresome.

**Comment by mabouels** [comment]
We are also hitting the exact same bug.

**Comment by thedrow** [comment]
It appears that by fixing #3921 we introduced this bug.
We cannot, however revert the fix since it will regress the other bug.
Suggestions are welcome :)

**Comment by gabrielmartin** [comment]
Okay, no problem. :)

One suggestion would be to change line 819 of kombu/connection.py from:
`self.ensure_connection()`
to:
`self.ensure_connection(max_retries=1, interval_start=0)`

This solves the problem when the given potential broker connection does not exist and prevents the function call from hanging. Although I do not know if such a change will affect the original issue you pointed out (#3921) or cause other problems.

For clarification, this issue is not limited to RabbitMQ, but also includes Redis (and probably any other broker type).

**Comment by georgepsarakis** [comment]
@gabrielmartin I am not sure to where exactly in the `kombu.connection` module you are referring, perhaps the `master` branch has been modified since. The closest line to 819, would be [here](https://github.com/celery/kombu/blob/master/kombu/connection.py#L831) and the `max_retries` is retrieved from transport options correctly. Perhaps I am missing something?

**Comment by gabrielmartin** [comment]
@georgepsarakis Hmm... I thought I had looked at the master branch in addition to the v4.1.0 tag, but maybe it changed recently as you said. Either way, it looks like this issue will be resolved in a future release of kombu.

Thank you!

**Comment by clokep** [comment]
@OllieBuilds and I are seeing a similar issue in my configuration, I don't think this is fixed for all cases. It looks like what @gabrielmartin pointed out was fixed in [kombu commit 0f4da8d45](https://github.com/celery/kombu/commit/0f4da8d45): transport broker options are now properly passed to the transport, so when installing kombu from git you're able to set your transport options to include a max retry count, e.g.:

```python
CELERY_BROKER_TRANSPORT_OPTIONS = {
    ‘max_retries’: 4,
}
```

What confuses me is that there's a [`broker_connection_max_retries`](http://docs.celeryproject.org/en/latest/userguide/configuration.html?highlight=task_publish_retry_policy#broker-connection-max-retries) option that seems like it should apply here too, but it's only used for the [worker's consumer](https://github.com/celery/celery/blob/120770929f4a37c5373a378b75b5c41a99702af9/celery/worker/consumer/consumer.py#L439).

It seems to me that either:
1. This is a documentation issue and `broker_connection_max_retries` should say that it applies only to workers. The [current documentation is](http://docs.celeryproject.org/en/latest/userguide/configuration.html?highlight=task_publish_retry_policy#broker-connection-max-retries):

> Maximum number of retries before we give up re-establishing a connection to the AMQP broker.

2. This is a bug that the producer doesn't also apply the `broker_connection_max_retries` setting when establishing a connection. (As far as I can tell, this seems to be somewhere near [`_connection` in the `Celery` class](https://github.com/celery/celery/blob/59ab9d93f4c570a135c62db992777aedbfede637/celery/app/base.py#L810-L835)?)

Any ideas how this is supposed to work? Thanks in advance for any thoughts!

**Comment by dralley** [comment]
@clokep , I also find that confusing.
